sw/host/opentitanlib/src/transport/chip_whisperer/spi.rs

```diff
@@ -86,6 +79,40 @@ impl<B: Board> Target for Spi<B> {
         Ok(true)
     }
 
+    fn set_pins(
+        &self,
+        serial_clock: Option<&Rc<dyn gpio::GpioPin>>,
+        host_out_device_in: Option<&Rc<dyn gpio::GpioPin>>,
+        host_in_device_out: Option<&Rc<dyn gpio::GpioPin>>,
+        chip_select: Option<&Rc<dyn gpio::GpioPin>>,
+    ) -> Result<()> {
+        match (
+            serial_clock,
+            host_out_device_in,
+            host_in_device_out,
+            chip_select,
+        ) {
+            (Some(clk), Some(sdo), Some(sdi), Some(cs)) => {
+                let usb = self.device.borrow();
+                usb.spi1_enable(false)?;
+                usb.spi1_setpins(
+                    sdo.get_internal_pin_name().unwrap(),
```

Comment:
What I really wanted, was a way for e.g. this SPI code of the Sam3X backend to be able to "recognize" a GPIO pin as belonging to the same Sam3X backend, and furthermore tell which pin it is, in whatever internal representation the backend uses.

I did not know how to do that in Rust, so I added the `fn get_internal_pin_name() -> Option<&str>` as the next best thing.  It does not allow verification that the comes from the same backend (that is, if I have a Rust program that simultaneously instantitates a Sam3X and a HyperDebug backend, it could pass a pin object belonging to one backend into the `set_pins()` method of the other, without detection), but it does allow for the backends I cared about (HyperDebug, Sam3X, Proxy) to tell which pin, in their internal representation.

Not all other backends use strings as the internal representation, most notably the UltraDebug and Dediprog backends use integer pin indices, and would not be able to implement a `fn get_internal_pin_name() -> &str` other than by panicking.  I guess that might not be worse than the current panic at every call site.

I was hoping, that if I were to change this interface, though, that it could be into something truly better, that would support verifying that it is the right kind of pin object.  Ideally, I would somehow be able to assert that the given `GpioPin` is of type `chip_whisperer::Pin<B>`, and furthermore, that its field `device: Rc<RefCell<Backend<B>>>` points at the same `Backend` instance as this `Spi<B>` object.  I do not know if that is possible in Rust, though.

Reply:
I get what you mean, I didn't spend much time thinking in a way of doing it in Rust, but it seems to me that it can be tackled using the type system with some kind of type state. But this may be a major refactor.

Reply:
Unfortunately, I think in Rust you cannot downcast an arbitrary trait, just Any, ie. given an instance of `dyn GpioPin`, you can't convert it to a `Pin` safely (you could do it with Any but then it becomes super ugly).
If you wanted some safety, you could add a method to `GpioPin` to return the device, to make sure that it comes from the correct backend, but that mean adding code to every backend.
Maybe @nbdd0121, @jwnrt or @HU90m  know a better way to achieve this?